Content edit to Fed 2 subgraph spec

StephenBarlow opened the pull request.

StephenBarlow commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,623 @@
+
+- Automatically extend its schema with all definitions listed in the [subgraph schema specification](#subgraph-schema-specification)
+- Correctly resolve the `Query._service` [enhanced introspection field](#fetch-service-capabilities)
+- Correctly generate [stubs for externally referenced types](#generating-stub-types)
```
Not sure this section needs to remain

StephenBarlow commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,623 @@
+Every subgraph must also automatically define the resolver for this field. The  logic for this resolver is as follows:
+
+1. For each entity representation included in the `representations` list, obtain the `__typename` from that representation.
+2. Execute the **reference resolver** for the corresponding `__typename` and provide the full representation to that resolver.
```
I suppose a "reference resolver" is how `@apollo/subgraph` happens to handle resolving representations, but it isn't the mandatory solution for other libraries, right? So I think we need to make this section a little more implementation-agnostic

dariuszkuc replied:
yes, libraries have to implement this logic

StephenBarlow commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,623 @@
+
+External schemas are identified by their `url`, which optionally ends with a name and version with the following format: `{NAME}`/v`{MAJOR}`.`{MINOR}`
+
+The presence of a `@link` directive makes a document a [core schema](https://specs.apollo.dev/#def-core-schema).
```
Is this outdated?

patrick91 replied:
maybe, the way link is declared in that spec is slightly different that how it is declared in federation:

```
directive @link(
  url: String!,  # url is non-nullable here
  as: String,
  import: [Import],
  for: Purpose
) repeatable on SCHEMA
```

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+scalar _Any
+scalar FieldSet
+scalar link__Import
```
technically `link__` namespace could be skipped IFF those elements are imported through `@link` directive, similarly how we skip the default `federation__` namespace for all other directives.

If elements are not imported through `@link` they need to be namespaced either with default value from the linked spec OR based on the specific namespace value (`@link(url: <spec url>, as: "myNamespace")`)

StephenBarlow replied:
Should the subgraph spec _require_ this definition alongside the other required definitions?

```
extend schema
  @link(url: "https://specs.apollo.dev/link/v1.0/", import: ["@link", "Import", "Purpose"])
```

Not necessarily because it's necessary to un-namespace `Import` and `Purpose`, but because otherwise `@link` isn't being tied to its associated spec

dariuszkuc replied:
I believe the current suggestion is to NOT import the link spec by default as it is automatically added by the router. Devs can import the link spec if they want to avoid name-spacing of `Import` scalar and `Purpose` enum, but should omit it otherwise.

Side note: when importing a spec, it automatically pulls in the corresponding directive matching the spec - so we don't need to explicitly import `@link` in the example above (by importing federation spec we also pull in `@federation` directive).

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+## Enhanced introspection with `Query._service`
+
+A federated graph router (also known as a gateway) can compose its supergraph schema dynamically at runtime. To do so, it first executes the following enhanced introspection query on each of its subgraphs to obtain all subgraph schemas:
```
AFAIK `_service { sdl }` query is only used for local composition (i.e. running in dev mode). Once gateway is deployed to test/prod we would use rover to publish the schemas.

StephenBarlow replied:
This is 100% what we _want_ folks to do, and ideally they wouldn't even use dynamic composition in dev mode. I'll add some bolding and prominence to the callout below to clarify just how much this is discouraged, but I suppose at least for now it is a core requirement for a subgraph.

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+* It must **include** all uses of all federation-specific directives, such as `@key`.
+    * For a list of all federation-specific subgraph directives, see the [glossary](#glossary-of-schema-additions).
+* It must **omit** all automatically added definitions from the [subgraph schema specification](#subgraph-schema-specification), such as `Query._service` and `_Service.sdl`!
```
this is true for fed v1, it is fine to include all the information in fed v2 schema

patrick91 replied:
What's our recommendation? Personally, I've left `_service` in strawberry, but maybe I should hide it?

What about `_entities`? Should that be removed as well?

dariuszkuc replied:
I believe we need to exclude it from `_service { sdl }` for fed 1 but it is fine to keep them for fed 2 (simpler as there is no need for custom logic to filter stuff)

patrick91 replied:
got it, maybe we should write that since it will simplify the implementations 😊

StephenBarlow replied:
Added a clarification that this is required only if supporting Fed 1

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+To support the enhanced introspection query, a subgraph service must define resolvers for the following fields:
+
+- `Query._service`
```
instead maybe provide this info as GraphQL SDL (as otherwise you will need to specify the nullability)?

```
extend type Query {
  _service: _Service!
}

type _Service {
  sdl: String!
}
```

dariuszkuc replied:
technically the only required resolver would be for `_entities` query if there is a federated type (type with `@key`) in the schema

StephenBarlow replied:
Good idea!

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+In the above example, the Shipping subgraph defines a `shippingCost` field for the `Product` entity. The `shippingCost` field `@requires` the value of the `weight` field, which is a field that's resolved by a _different_ subgraph. To indicate that `weight` is resolved only by _other_ subgraphs, the Shipping subgraph _also_ defines the field, but it applies the `@external` directive to it.
+
+This allows us to be explicit about dependencies on another service, and service composition will verify that an `@external` field matches up with the original field definition, which can catch mistakes or migration issues (when the original field changes its type for example). `@external` fields also give an individual service the type information it needs to validate and decode incoming represe ...
```
nit: in fed v1 `@external` was required on any field referenced from `@key/@requires/@provides`, fed v2 relaxed this requirement and now it is only needed for fields referenced by `@requires` directive and in some cases for `@provides` (only needed if field is resolved locally in some but not all paths)

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+Each subgraph in a supergraph should be able to run successfully in isolation, regardless of whether any _other_ subgraphs are running. Many subgraphs run successfully in isolation with no additional logic required. However, subgraph fields _can_ depend on types and fields that are defined in _other_ subgraphs. Whenever this is the case, a subgraph needs to define its own **stubs** for these exte ...
+
+To indicate a dependency on another subgraph's type or field, a subgraph schema field uses either the `@requires` or `@provides` directive:
```
maybe we should also add an example for `@provides`?

StephenBarlow replied:
I'm going to leave it to one example in the federated directives article, just because the listings for `@requires` and `@provides` are both very close by and provide their own examples

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+Each item in the `representations` list is an `_Any` scalar. This is a federation-specific scalar defined in the [subgraph schema specification](#subgraph-schema-specification). This generic scalar enables the graph router to include representations of _different_ entities in the _same_ query, all of which can have a different shape.
+
+The `Query._entities` field returns a list of objects that belong to the `_Entity` union type, which is why this union must include _all_ entities that the subgraph defines.
```
lets call out the order -> `_entities` return a list of objects that are in the same order as the passed in representations (there was a question about it in the past)

StephenBarlow replied:
good call

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+The graph router uses this entry point to directly fetch fields of entity objects. It combines those fields with _other_ fields of the same entity that are returned by _other_ subgraphs.
+
+The `Query._entities` field takes a required `representations` argument, which is a list of entity **representations**. A representation is an object that contains all of an entity's `@key` fields, plus its `__typename` field. These are the fields that a subgraph can use to uniquely identify a particular instance of an entity.
```
nit: entity can specify multiple `@key`s (it is a repeatable directive) so it will contain one of field sets that identity the target object

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+This object type must have an `sdl: String!` field, which returns the SDL of the subgraph schema as a string.
+
+The returned schema string _must_ include all uses of federation-specific directives (`@key`, `@requires`, etc.), and it _must not_ include any definitions from the [subgraph schema specification](#subgraph-schema-specification).
```
again this is true for fed v1, fed v2 can include full SDL definition with all the types

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+### `scalar _Any`
+
+This scalar is the type used for entity **representations** that the graph router passes to the `Query._entities` field. An `_Any` scalar is validated by matching its `__typename` and `@key` fields against entities defined in the subgraph schema.
```
do we need to call out this is basically a map of field names/field value pairs (values could be scalars or objects)?

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+### `scalar FieldSet`
+
+This string-serialized scalar represents a set of fields that's passed to a federated directive, such as `@requires` or `@provides`.
```
I'd call out `@key` as that is the primary use case

dariuszkuc commented on docs/source/federation-spec.mdx:
````diff
@@ -0,0 +1,625 @@
+directive @external on FIELD_DEFINITION
+```
+
+The `@external` directive indicates that a particular subgraph _cannot_ resolve a particular entity field. This enables a subgraph to reference a _different_ subgraph's field _without_ also resolving that field.
````
this is not completely true

```
type Foo @key("id") {
  id: ID!
  localBar: Bar @provides("baz") # resolves field locally even though it is marked as external
  remoteBar: Bar # resolves field remotely
}

extend type Bar @key("id) {
  id: ID!
  baz: String @external
}
```

dariuszkuc replied:
in fed v2 `@external` is only required for `@required` fields and in some cases (see above) `@provides` fields

StephenBarlow replied:
Right, I use better language for this in the other directives article and I'll make sure to review. The point being at least that a subgraph _usually_ can't resolve an `@external` field unless there's a path that allows it via `@provides`

dariuszkuc commented on docs/source/federation-spec.mdx:
````diff
@@ -0,0 +1,625 @@
+directive @inaccessible(from: String!) on FIELD_DEFINITION | INTERFACE | OBJECT | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION
+```
+
+The `@inaccessible` directive indicates that a location in the subgraph schema should _not_ be included in the [API schema](./federated-types/overview/#api-schema) that clients use to execute operations against the graph router.
````
lets add a primary use case -> it can be used for defining private keys that cannot be queried by the api schema but will be used by gateway for composition

StephenBarlow replied:
This is a cool and powerful use case that is very important and that I had never considered 😮

dariuszkuc commented on docs/source/federation-spec.mdx:
````diff
@@ -0,0 +1,625 @@
+### `@key`
+
+```graphql
+directive @key(fields: FieldSet!, resolvable: Boolean = true) repeatable on OBJECT | INTERFACE
````
lets add some docs about the `resolvable = false`

dariuszkuc commented on docs/source/federation-spec.mdx:
````diff
@@ -0,0 +1,625 @@
+directive @override(from: String!) on FIELD_DEFINITION
+```
+
+The `@override` directive indicates that the current subgraph takes responsibility for resolving a field _away_ from the subgraph specified in the `from` argument.
````
do we need to call out that the subgraph name has to match the one in the studio?

dariuszkuc commented on docs/source/federation-spec.mdx:
````diff
@@ -0,0 +1,625 @@
+}
+```
+
+By default, this subgraph _can't_ resolve the `Product.upc` or `Product.name` fields, so it marks those fields as `@external`. However, this subgraph _can_ resolve the `Product.name` field specifically when it's a subfield of `Review.product`. So, it applies the `@provides` directive to `Review.product` to indicate this fact to the graph router.
````
technically federation v2 does not require `@external` on keys anymore, similarly if that is the only query path that `Product.name` can be resolved then there is no need for `@external` neither

StephenBarlow replied:
Good call, this content is now cut so no problem here

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION
+directive @override(from: String!) on FIELD_DEFINITION
+
+# This is an optional directive (see the glossary below)
```
where is the glossary?

StephenBarlow replied:
Replaced this comment to be more standalone and added `@extends` to the federated directives article

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+Grammatically, a `FieldSet` is a [selection set](http://spec.graphql.org/draft/#sec-Selection-Sets) minus the outermost curly braces. It can represent a single field (`"upc"`), multiple fields (`"id countryCode"`), and even nested selection sets (`"id organization { id }"`).
+
+### == Added directives ==
```
should we also list `@extends`? it is used by some libs as a workaround for lack of the `extend type` syntax

StephenBarlow replied:
Yes, I'll get this back in—I prematurely cut it without asking purely because `extend` became much less important of a keyword for Fed 2, but there are certainly still applications for it

patrick91 commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+  _service: _Service!
+}
+
+directive @external on FIELD_DEFINITION
+directive @requires(fields: FieldSet!) on FIELD_DEFINITION
+directive @provides(fields: FieldSet!) on FIELD_DEFINITION
+directive @key(fields: FieldSet!, resolvable: Boolean = true) repeatable on OBJECT | INTERFACE
+directive @link(url: String, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA
+directive @shareable on OBJECT | FIELD_DEFINITION
+directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION
+directive @override(from: String!) on FIELD_DEFINITION
```
My understanding was that printing/writing the directive definition inside schema was optional, is that not the case?

dariuszkuc replied:
yes including them in printed SDL (returned by `_service { sdl }` query) is optional but I think they have to be defined within the GraphQL server in order to run (at least `graphql-java` requires definitions to be present)

patrick91 replied:
ah, that makes sense for schema first libraries!

StephenBarlow replied:
Added language to clarify that these definitions need to exist within the server via whatever mechanism the library supports (if not via static SDL)

patrick91 commented on docs/source/federation-spec.mdx:
````diff
@@ -0,0 +1,625 @@
+}
+```
+
+Every subgraph must also automatically define the resolver for this field. The  logic for this resolver is as follows:
````
```suggestion
Every subgraph must also automatically define the resolver for this field. The logic for this resolver is as follows:
```

patrick91 commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,625 @@
+
+The presence of a `@link` directive makes a document a [core schema](https://specs.apollo.dev/#def-core-schema).
+
+The `for` argument describes the purpose of a `@link`. Currently accepted values are `SECURITY` or `EXECUTION`. Core schema-aware servers such as Apollo Router and Gateway will refuse to operate on schemas that contain `@link`s to unsupported specs which are `for: SECURITY` or `for: EXECUTION`.
```
> which **are not**

?

StephenBarlow commented:
@dariuszkuc @patrick91 Thanks for all the thoughtful feedback! I've updated the PR with one squashed commit-of-commits, so the diff since the last review should all be straightforward here: https://github.com/apollographql/federation/pull/2045/commits/67d774314989ec3ed7481d6a9414e160f9c068f5

Hopefully this properly addresses the majority of outstanding feedback and we can knock out any remaining concerns this week.

dariuszkuc commented on docs/source/federated-types/federated-directives.mdx:
```diff
@@ -32,13 +32,30 @@ extend schema
 This example subgraph schema uses `@primaryKey` for the federated directive usually named [`@key`](#key).
 
-## Managing entities
+### `@link`
```
its somewhat odd that `@link` definition appears after mentioning it above for importing and renaming the directives - maybe it should be moved up?

dariuszkuc replied:
also should we add sections about namespacing? i.e. if you don't explicitly import types then they should be namespaced with `@federation__` prefix (unless custom one is provided)

StephenBarlow replied:
I'd prefer to leave the `@link` section below importing and renaming, because those sections are critical setup for making use of any other directive on the page, whereas the `@link` section is more reference like the rest of the directives.

Yes agreed on the namespacing, I'll add that

dariuszkuc commented on docs/source/federated-types/federated-directives.mdx:
````diff
@@ -48,6 +65,20 @@ type Product @key(fields: "id") {
+}
+```
+
+To check whether your subgraph library supports repeatable directives, see the `@KEY (MULTI)` item in [Federation-compatible subgraph implementations](../supported-subgraphs/).
````
actually we just fixed that in the subgraph repo and added a proper test (previously we were not really checking for it) -> new test is named `repeatable @key` (we still got ~half of implementations to update to the latest schema, once updated we'll fix the docs)

StephenBarlow replied:
Ok! Should I just cut this for now and we can add a corresponding sentence when update is complete?

dariuszkuc replied:
lets change it to the expected `repeatable @key` test as otherwise pretty sure we'll forgot about it (`@key multi` is for checking keys that reference multiple fields). I can partially update the results.

dariuszkuc commented on docs/source/federated-types/federated-directives.mdx:
```diff
@@ -48,6 +65,20 @@ type Product @key(fields: "id") {
+
+To check whether your subgraph library supports repeatable directives, see the `@KEY (MULTI)` item in [Federation-compatible subgraph implementations](../supported-subgraphs/).
+
+> Although it's valid to apply `@key` to an interface definition, this currently has no effect in Apollo Federation.
```
I thought that it would be inherited by the implementing classes? so if you add `@key` to the interface then all implementing classes will have it applied automatically

StephenBarlow replied:
Ooh maybe in Fed 2? Last I knew in Fed 1 this did nothing but was valid syntax. I could be out of the loop though

StephenBarlow replied:
Okay, looks like Fed 2 composition currently straight-up _rejects_ `@key` on interfaces despite being valid syntax, so I'll update this accordingly

dariuszkuc commented on docs/source/federated-types/federated-directives.mdx:
```diff
@@ -102,13 +133,31 @@ The default value is `true`.
+
+Indicates that an object or interface definition is an extension of _another_ definition of that same type.
+
+> ⚠️ **If you define your subgraph schema in a static text file, do not use this directive!** Instead, use GraphQL's built-in `extend` keyword.
```
unsure if this is true, I think the general rule is -> use `@extends` only if your subgraph library does not support the `extend` keyword syntax

StephenBarlow replied:
Cool, tweaked the language here

dariuszkuc commented on docs/source/federation-spec.mdx:
```diff
@@ -0,0 +1,447 @@
+}
+
+type _Service {
+  sdl: String
```
```suggestion
  sdl: String!
```

StephenBarlow commented:
@dariuszkuc Pushed a followup commit that hopefully addresses remaining feedback! https://github.com/apollographql/federation/pull/2045/commits/d150fadf9e607e44e9df2c765d42de8f266d5ee9

It also folds in `@composeDirective`, which was added to the subgraph spec while this PR has been open.

Does this PR look okay in terms of "no regressions from live docs" at this point? If so, might be nice to get it out so any followup changes feel more manageable than this now-hefty PR

dariuszkuc reviewed (approved):
Awesome work!
